include/rsp_dma.inc

```diff
@@ -0,0 +1,183 @@
+#     to write direction-agnostic transfer code that can be then invoked
+#     passing in the direction in a register.
+#
+# The RSP DMA controller allows to enqueue one transfer while another one
+# is in progress. You can use the *Async functions to do this. If you try
+# to enqueue a third transfer, though, the *Async function will automatically
+# block waiting for the first one to finish. This allows the code to call
+# *Async functions freely and get "best effort" background transfers.
+# 
+# When you need to explicitly synchronize (eg: to access read data), you can
+# use DMAWaitReady to wait until the DMA engine is ready to receive another
+# transfer, or DMAWaitIdle to wait until all DMA transfers are finished. See
+# their documentation below.
+#
+# The DMA controller is able to read/write "rectangular portions"
+# of memory in RDRAM (that is, you can define width, height and
+# pitch to extract a portion of a larger image). Accesses in DMEM
+# are always linear, though. If you don't need this feature,
+# you can set the transfer size as width and use 1 as height: this way,
+# the pitch will be ignored.
+#
+# At the hardware level, both the DMEM and the RDRAM addresses are
+# treated as 8-byte aligned (that is, the lowest 3 bits are ignored).
+# If you issue a transfer from RDRAM with an unaligned address (in s0),
+# the DMEM address (s4) will be updated to reflect where the byte that
+# was actually requested ended up. For instance, if s0 = 0x80041203, s4=0x100,
+# after the function s4 will be 0x103, as that is the address where the
+# byte at 0x80041203 ended up being written to.
+#
+# INPUT:
+#   t0: transfer size as width/height (use DMA_SIZE macro)
+#   t1: pitch (ignored if height==1)
+#   s0: RDRAM address
+#   s4: DMEM address. After DMA, updated with unalignment
+#
+#   Only for DMAExec:
+#   t2: DMA operation flags: select direction and sync/async.
+#       Use one of the DMA_* macros below. If async, you can call
+#       DMAWaitIdle to synchronize later if required.
+#
+# OUTPUT:
+#   Only for DMAIn:
+#   s4: adjusted to point to the first requested byte
+#       (because of DMA alignment, more bytes might have been
+#       DMA'd).
+#
+# DESTROY:
+#   t2, at
+# 
+#########################################################
+
+    .func DMAOut
+DMAOut:
+    j DMAExec
+    li t2, DMA_OUT
+    .endfunc
+
+    .func DMAOutAsync
+DMAOutAsync:
+    j DMAExec
+    li t2, DMA_OUT_ASYNC
+    .endfunc
+
+    .func DMAInAsync
+DMAInAsync:
+    j DMAExec
+    li t2, DMA_IN_ASYNC
+    .endfunc
+
+    .func DMAIn
+DMAIn:
+    li t2, DMA_IN
+    # fallthrough!
+    .endfunc
+
+    .set noat
+    .func DMAExec
+DMAExec:
+    # Wait for the DMA engine to be ready. We do this
+    # every time because with async transfers it's easy to get
+    # confused and forget to wait.
+    mfc0 $1, COP0_DMA_FULL
+    bnez $1, DMAExec
+    # Merge pitch into the width/height word, converting it
+    # into "skip" (= pitch-width)
+    andi $1, t0, 0xFFFF
```

Comment:
Wasnt the [width](https://github.com/DragonMinded/libdragon/pull/166/files#diff-0b1db597974243386a372fc3028af955f77db23942d78955b06a2bbeea8a5d6dR665) 12bits?

Reply:
Good catch!

It looks like I introduced this bug in Jun20 in my codebase, when I introduced async functions, with this diff:

```diff
-    andi $1, t0, 0xFFF
+    .func DMAInOut
+DMAInOut:
+    mfc0 $1, COP0_SP_STATUS
+    andi $1, SP_STATUS_DMA_FULL
+    bnez $1, DMAInOut
+    andi $1, t0, 0xFFFF
```

I have never realized it because it happens not to matter... that value is used to compute the "skip" component which is only 12-bit wide, so those last 4 bits do not affect the result and they are shifted away later with `sll $1, 20`. Nonetheless, I fixed it. Thanks!